### PR TITLE
CG-324 - Update gateway-ethereum-ts to use sendable transaction type

### PR DESCRIPTION
**Pull request by TYRONEMICHAEL**
This PR exposes an options property that, if set, will wrap the Gateway Transaction in a Meta Transaction and send via the Forwarder contract. It also exposes a new Sendable Transaction Data Type and Sent Transaction Type that we can now pass back to the client for the client to send the transaction and pay the gas.

### CODE REVIEW

File: ethereum/gateway-eth-ts/examples/eth-ts-calldata.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ require("dotenv/config");
     await gtCallDataLib.init();
     const testUser = '0x57AB42d4fa756b6956b0cAf986a5f53bA90D9e28';
     
-    let gatewayToken = await gtCallDataLib.gatewayTokens["0xa3894BbA27f4Be571fFA319D02c122E021024cF2"].tokenInstance
+    let gatewayToken = await gtCallDataLib.gatewayTokens["0x67306284Fb127E9baF713Ebf793d741cE763F81A"].tokenInstance
```
Comment (dankelleher) [inline ethereum/gateway-eth-ts/examples/eth-ts-calldata.js:18]: Can you make this a constant in a shared file for all examples?

File: ethereum/gateway-eth-ts/src/GatewayTs.ts [hunk 2]
```diff
@@ -45,19 +47,41 @@ export class GatewayTs extends GatewayTsBase {
...
+
+    if (options?.forwardTransaction) {
+      const { request, signature } = await signMetaTxRequest(this.wallet, this.forwarder.contract, {
+        from: this.wallet.address,
+        to: gatewayTokenContract.contract.address,
+        data: txRequest.data
+      });
+
+      const metaTx = await populateTx(this.forwarder.contract, 'execute', [request, signature], options);
+      const sendableTransaction = new SendableTransaction(gatewayTokenContract.contract, metaTx, options);
```
Comment (dankelleher) [inline ethereum/gateway-eth-ts/src/GatewayTs.ts:79]: line duplication here between 78,79 and 83, 84. You can just create the tx/metatx and  then pass it into the final line which should just be return new SendableTransaction...:

So something like

```
const txRequest = await populateTx(...)
const txToSend = wrapTxIfForwarded(txRequest, options, ...)
return new SendableTransaction(..., txToSend,...)
```

Note also, this pattern applies to _all transactions_, not just issue, so the wrapTxIfForwarded (or whatever you choose to name it) should not be specific to issue. The logic for signing and calling the forwarder should not be in issue().

**Review by dankelleher — APPROVED** [review]

File: ethereum/gateway-eth-ts/src/GatewayTsCallData.test.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ describe("Test GatewayTSCallData class", function () {
   const defaultGas: number | BigNumber = 6_000_000;
   const defaultGasPrice: number | BigNumber = 1_000_000_000_000;
 
-  const sampleWalletAddress = "0x57AB42d4fa756b6956b0cAf986a5f53bA90D9e28";
+  const sampleWalletAddress = "0xD42Ef952F2EA1E77a8b771884f15Bf20e35cF85f";
```
Comment (dankelleher) [inline ethereum/gateway-eth-ts/src/GatewayTsCallData.test.ts:37]: should this be in some config file?

**Review by dankelleher — APPROVED** [review]

**Review by dankelleher — APPROVED** [review]

File: ethereum/gateway-eth-ts/examples/eth-ts-calldata.js (hunk 1 above)
Comment (rado0x54) [reply ethereum/gateway-eth-ts/examples/eth-ts-calldata.js:18]: Agreed. This should just come from the lib as well.

File: ethereum/gateway-eth-ts/src/GatewayTs.ts [hunk 4]
```diff
@@ -79,103 +81,191 @@ export class GatewayTs extends GatewayTsBase {
...
     gatewayTokenAddress?: string,
     options?: TxOptions
-  ): Promise<string> {
-    const gatewayToken = this.getGatewayTokenContract(gatewayTokenAddress);
+  ): Promise<SendableTransaction> {
+    const gatewayTokenContract =
+      this.getGatewayTokenContract(gatewayTokenAddress);
+    const { contract } = gatewayTokenContract;
 
     if (tokenId === null) {
```
[the remark this answers is not in the document]
Comment (rado0x54) [reply ethereum/gateway-eth-ts/src/GatewayTs.ts:89]: Ok I see that null is the default type in the function, so I guess it does not need to be done in this PR.

File: ethereum/gateway-eth-ts/src/GatewayTsBase.ts [hunk 5]
```diff
@@ -100,7 +107,7 @@ export class GatewayTsBase {
     let gatewayToken: GatewayToken;
 
     if (gatewayTokenAddress) {
-      gatewayToken = this.gatewayTokens[gatewayTokenAddress].tokenInstance;
+      gatewayToken = this.gatewayTokens[gatewayTokenAddress]?.tokenInstance;
```
Comment (rado0x54) [inline ethereum/gateway-eth-ts/src/GatewayTsBase.ts:110]: why was the questionmark added here? Not sure if the function is robust to returning `undefined`.

File: ethereum/gateway-eth-ts/examples/eth-ts-calldata.js (hunk 1 above)
Comment (chriteixeira) [reply ethereum/gateway-eth-ts/examples/eth-ts-calldata.js:18]: This is an example file that is written with JS. The other occurrences on TS code were fixed.

File: ethereum/gateway-eth-ts/src/GatewayTsBase.test.ts [hunk 6]
```diff
@@ -25,7 +24,7 @@ describe("Test GatewayTSBase class", function () {
   let defaultGatewayToken: string;
   const defaultGas: number | BigNumber = 6_000_000;
 
-  const sampleWalletAddress = "0x57AB42d4fa756b6956b0cAf986a5f53bA90D9e28";
+  const sampleWalletAddress = "0xD42Ef952F2EA1E77a8b771884f15Bf20e35cF85f";
```
Comment (rado0x54) [inline ethereum/gateway-eth-ts/src/GatewayTsBase.test.ts:27]: Export to config.

File: ethereum/gateway-eth-ts/src/contracts/Forwarder.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  constructor(signerOrProvider: Signer | BaseProvider, addressOrName: string) {
+    this.contract = new Contract(
+      addressOrName,
+      abis.Forwarder,
+      signerOrProvider
+    );
+  }
+
+  execute = (request: EIP712Message, signature: string) => {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-return
```
Comment (rado0x54) [inline ethereum/gateway-eth-ts/src/contracts/Forwarder.ts:18]: are we able to type this more (similar to the other contracts?)

File: ethereum/gateway-eth-ts/src/utils/signer.ts [hunk 8]
```diff
@@ -44,3 +45,113 @@ export const mnemonicSigner = function (
...
+  primaryType: "ForwardRequest",
+});
+
+async function signTypedData(signer: Wallet, data: EIP712TypedData) {
+  const types = { ForwardRequest } as Record<string, Array<TypedDataField>>;
+  return signer._signTypedData(data.domain, types, data.message);
+}
+
+const buildRequest = async (
+  forwarder: Contract,
```
Comment (rado0x54) [inline ethereum/gateway-eth-ts/src/utils/signer.ts:127]: Can we not build an Interface for the forwarder Contract (as with the other contracts), to make the TS a little nicer?

**Review by rado0x54 — APPROVED** [review]

**Pull request merged**